precompiles/erc20/integration_test.go

```diff
@@ -0,0 +1,1199 @@
+				// only support the actual token denomination and don't know of other balances.
+			)
+
+			DescribeTable("it should return zero if the account does not exist", func(callType int) {
+				address := utiltx.GenerateAddress()
+
+				// Query the balance
+				txArgs, balancesArgs := s.getTxAndCallArgs(callType, contractData, erc20.BalanceOfMethod, address)
+
+				_, ethRes, err := s.factory.CallContractAndCheckLogs(sender.Priv, txArgs, balancesArgs, passCheck)
+				Expect(err).ToNot(HaveOccurred(), "unexpected result calling contract")
+
+				var balance *big.Int
+				err = s.precompile.UnpackIntoInterface(&balance, erc20.BalanceOfMethod, ethRes.Ret)
+				Expect(err).ToNot(HaveOccurred(), "failed to unpack result")
+				Expect(balance.Int64()).To(BeZero(), "expected zero balance")
+			},
+				Entry(" - direct call", directCall),
+				Entry(" - through contract", contractCall),
+				Entry(" - through erc20 contract", erc20Call),
+			)
+		})
+
+		When("querying allowance", func() {
+			DescribeTable("it should return an existing allowance", func(callType int) {
+				grantee := utiltx.GenerateAddress()
+				granter := sender
+				authzCoins := sdk.Coins{sdk.NewInt64Coin(s.tokenDenom, 100)}
+
+				// Make topics for the grantee and granter common.Address types
+				granteeTopic, err := cmn.MakeTopic(grantee)
+				Expect(err).ToNot(HaveOccurred(), "failed to make topic")
+				println("Grantee topic: ", granteeTopic.String())
+				granterTopic, err := cmn.MakeTopic(granter.Addr)
+				Expect(err).ToNot(HaveOccurred(), "failed to make topic")
+				println("Granter topic: ", granterTopic.String())
+				erc20Topic, err := cmn.MakeTopic(contractData.erc20Addr)
+				Expect(err).ToNot(HaveOccurred(), "failed to make topic")
+				println("ERC20 contract topic: ", erc20Topic.String())
+				ownerTopic, err := cmn.MakeTopic(sender.Addr)
+				Expect(err).ToNot(HaveOccurred(), "failed to make topic")
+				println("Owner topic: ", ownerTopic.String())
+
+				s.setupSendAuthzForContract(callType, contractData, grantee, granter.Priv, authzCoins)
+
+				txArgs, allowanceArgs := s.getTxAndCallArgs(callType, contractData, auth.AllowanceMethod, granter.Addr, grantee)
+
+				_, ethRes, err := s.factory.CallContractAndCheckLogs(granter.Priv, txArgs, allowanceArgs, passCheck)
+				Expect(err).ToNot(HaveOccurred(), "unexpected result calling contract")
+
+				var allowance *big.Int
+				err = s.precompile.UnpackIntoInterface(&allowance, auth.AllowanceMethod, ethRes.Ret)
+				Expect(err).ToNot(HaveOccurred(), "failed to unpack result")
+				Expect(allowance).To(Equal(authzCoins[0].Amount.BigInt()), "expected different allowance")
+			},
+				Entry(" - direct call", directCall),
+				Entry(" - through contract", contractCall),
+				// FIXME: This returns an empty allowance but the logs show that an approval was made - however the approval was made for a different owner??
+				Entry(" - through erc20 contract", erc20Call),
```

Comment:
This is very curious and I'm currently trying to figure out what's wrong here. When calling this using the ERC20MinterBurner contract, I'm printing the transaction logs and can see that there is indeed an approval made during `setupSendAuthzForContract`.

This function is used to send the call to `approve` on the ERC20 contract. However, when printing the inputs and the logs it's obvious that the approval is made to an unknown address as the owner?? It's neither the granter address nor the ERC20 address, yet the spender/grantee is set correctly.. Not too sure what to make of it right now, the OpenZeppelin contracts should just pass the message sender (which would be the granter) to the approve method (see [implementation](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/v4.9.3/contracts/token/ERC20/ERC20.sol#L126-L140)).

Printing some logs here:
```
values at time of input
---------------------------
Grantee topic:        0x000000000000000000000000b0762438fecccfa80adf1e168f2a3b44ccae3425
Granter topic:        0x000000000000000000000000336c9b4f79c53137508cd3c0d60cdb47909576a0
ERC20 contract topic: 0x0000000000000000000000003f6f31fed116814c088b87adae2d46c444a647b2
Owner topic:          0x000000000000000000000000336c9b4f79c53137508cd3c0d60cdb47909576a0

values in approval event log 
---------------------------
owner:    0x0000000000000000000000001f0991821904b62758dece69ba18e45ca5ae48c1
spender:  0x000000000000000000000000b0762438fecccfa80adf1e168f2a3b44ccae3425
```

Reply:
Checked this using Remix and it works as expected, so still debugging here..

Reply:
Okay found the problem, applying the fix now.